llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -5838,9 +5840,21 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
     return SDValue();
   }
   case Intrinsic::experimental_cttz_elts: {
-    SDValue NewCttzElts =
-        DAG.getNode(AArch64ISD::CTTZ_ELTS, dl, MVT::i64, Op.getOperand(1));
+    SDValue CttzOp = Op.getOperand(1);
+    EVT VT = CttzOp.getValueType();
+
+    if (!VT.isScalableVector()) {
+      // Retrieve original fixed-width vector from ISD::TRUNCATE Node.
+      assert(CttzOp.getOpcode() == ISD::TRUNCATE && "Expected ISD::TRUNCATE!");
```

Comment:
This is not a safe assumption to make.  Also, "masks" have a special format as dictated by `getBooleanContents` which `convertFixedMaskToScalableVector` is allowed to depend on.

Putting this together I think you should emit a suitable compare against zero to construct a mask that can be passed to `convertFixedMaskToScalableVector`.  I'd expect DAG combine to do a good job of removing this compare for the truncate scenario your expecting.

Reply:
Sorry, I hit the wrong button and thus posted my review comments early whilst still investigating the code.

I now understand where the truncate comes from but as before, this is something that cannot be relied upon.  For AArch64 `getBooleanContents` requires fixed length masks to be all zeros or all ones.  Given the change to `shouldExpandCttzElements` we can be sure the operand type will be an i1 vector (which you can and should assert for) and thus I think you can just sign extend to the relevant type (see `getTypeToTransformTo`) and pass this to `convertFixedMaskToScalableVector`.  As before, I'd expect this sign extend will be optimised away.